manager/main.go

```diff
@@ -62,34 +57,62 @@ func init() {
 	//+kubebuilder:scaffold:scheme
 }
 
-func initFlags(fs *pflag.FlagSet) {
-	fs.StringVar(&metricsAddr, "metrics-bind-address", "localhost:8080", "The address the metric endpoint binds to.")
-	fs.StringVar(&probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
-	fs.BoolVar(&enableLeaderElection, "leader-elect", false,
+type config struct {
+	metricsAddr          string
+	enableLeaderElection bool
+	probeAddr            string
+	gates                []string
+	logging              *logsv1.LoggingConfiguration
+}
+
+func newConfig() *config {
+	c := &config{
+		logging: logsv1.NewLoggingConfiguration(),
+	}
+	c.logging.Format = logsv1.JSONLogFormat
+	c.logging.Verbosity = logsv1.VerbosityLevel(4)
+
+	return c
+}
+
+func initFlags(fs *pflag.FlagSet, config *config) {
+	logsv1.AddFlags(config.logging, fs)
+
+	fs.StringVar(&config.metricsAddr, "metrics-bind-address", "localhost:8080", "The address the metric endpoint binds to.")
+	fs.StringVar(&config.probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
+	fs.BoolVar(&config.enableLeaderElection, "leader-elect", false,
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
-	fs.StringSliceVar(&gates, "feature-gates", []string{}, "A set of key=value pairs that describe feature gates for alpha/experimental features. ")
+	fs.StringSliceVar(&config.gates, "feature-gates", []string{}, "A set of key=value pairs that describe feature gates for alpha/experimental features. ")
 }
 
 func main() {
-	opts := kzap.Options{
-		Development: true,
-		Level:       zap.NewAtomicLevelAt(-4),
-	}
-	opts.BindFlags(flag.CommandLine)
+	config := newConfig()
 	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
-
-	initFlags(pflag.CommandLine)
+	initFlags(pflag.CommandLine, config)
 	pflag.Parse()
-	ctrl.SetLogger(kzap.New(kzap.UseFlagOptions(&opts)))
-	features.FeedGates(gates)
+
+	// Temporary logger for initialization
+	setupLog := ctrl.Log.WithName("setup")
+
+	if err := logsv1.ValidateAndApply(config.logging, nil); err != nil {
+		setupLog.Error(err, "unable to start manager")
+		os.Exit(1)
+	}
+
+	// klog.Background will automatically use the right logger.
+	ctrl.SetLogger(klog.Background())
+	// Once controller-runtime logger has been setup correctly, retrieve again
+	setupLog = ctrl.Log.WithName("setup")
```

Comment:
What's the purpose of defining this again? Not aware if there is some caching logic here that pulls the same logger with that name, which I see is passed onto all the reconcilers too.

Reply:
I'm not sure I understand the question about the cache
The purpose is more or less what the comment says: at the beginning of the execution, there is a default logger. Once the configuration has been read and applied, a new logger with the proper setup is created and set with `ctrl.SetLogger`. This line just retrieves that new configured logger to use for the rest of the setup execution.

Reply:
I see the same line on line 96 above so I was wondering how we were modifying `ctrl.Log` and retrieving a different instance of `setupLog`, so I was wondering what might have changed and whether this log is being stored somewhere.

Reply:
`ctrl.Log` gets modified when we call `ctrl.SetLogger`
Which we set with the klog logger. Which is configured when we call `VerifyAndApply`